docs/architecture/adr-009-transport-agnostic-peer-abstraction.md

```diff
@@ -0,0 +1,194 @@
+# ADR 009: Transport agnostic Peer abstraction
+
+## Changelog
+* 2021-02-05: drafted
+
+## Context
+
+With the opportunity to design and implement the peer-to-peer stack from
+scratch in the context of the Tendermint implementation in Rust, a lot of the
+learnings of the shortcomings of the original Go implementation can be used to
+prevent certain mistakes. Namely two:
+
+* Leakage of physical concerns into the core domain
+* Flexibility to adopt different wire protocols for transport of messages
+
+For that, the first set of newly introduced concepts will attempt to be generic
+over which transport is used to connect and converse with other peers. Given
+strongly tailored abstract interfaces, concrete implementations will be easy to
+spin up and plug into the machinery which lifts bytes from the wire into the
+core domain and transports messages into the rest of the system.
+
+## Decision
+
+### Transport
+
+Wraping the design is the `Transport`. Modelled with the properties of
+a physical network endpoint in mind, which can be bound and stopped. It should
+strongly correspond to the acquisition and lifecycle management of network
+resources on the system.
+
+``` rust
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;
+
+    fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;
+    fn shutdown(&self) -> Result<()>;
+}
+```
+
+After the successful bind the caller holds an `Endpoint` as well as a stream of
+incoming `Connection`s. Which is a standardised way to connect to new peers and
+react to newly connected ones respectively.
+
+``` rust
+pub trait Endpoint: Send {
+    type Connection;
+
+    fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;
+    fn listen_addrs(&self) -> Vec<SocketAddr>;
```

Comment:
1) can we add a comment here and for all other functions as well?

2) can we move this into `Incoming`? it makes more sense since the iterator (TcpListener.incoming() in case of MConnection) is being used there.

Reply:
> can we add a comment here and for all other functions as well?

Will be addressed together with all other clippy warnings, once it's final.

> can we move this into `Incoming`? it makes more sense since the iterator (TcpListener.incoming() in case of MConnection) is being used there.

Not quite getting the meaning here. `Incoming` is an iterator, not a trait that is implemented.

Reply:
> Not quite getting the meaning here. Incoming is an iterator, not a trait that is implemented.

please ignore